File: javascript/euler4.js
'use strict';

function isPalindrome(str) {
  return str === str.split('').reverse().join('');
}

var largestPalindrome = 0;
for(var i = 100; i < 1000; i++) {
  for(var j = 100; j < 1000; j++) {
    var num = i * j;
    if (isPalindrome('' + num)) {
       largestPalindrome = Math.max(largestPalindrome, num);
    }
  }
}
console.log(largestPalindrome);

File: clojure/euler4.clj
(defn ispalindromstr? [^long n]
  (let [s (str n)]
    (= (seq s) (reverse s))
  )
)

(defn ispalindromenum? [^long num]
  (loop [rev 0
         temp num]
    (if (> temp 0)
      (recur (+ (* rev 10) (mod temp 10)) (quot temp 10))
      (= rev num)
    )
  )
)

(defn euler4 [palifunc]
  (reduce max
    (for [i (range 999 0 -1) 
          j (range 999 0 -1)
          :let [num (* i j)]
          :when (palifunc num)]
      num
    )
  )
)

(println (euler4 ispalindromstr?)) ; with string based palifunc
(println (euler4 ispalindromenum?)) ; with num based palifunc
